server/src/main/java/io/crate/execution/engine/aggregation/impl/NumericSumAggregation.java

```diff
@@ -0,0 +1,330 @@
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.engine.aggregation.impl;
+
+import io.crate.breaker.RamAccounting;
+import io.crate.common.annotations.VisibleForTesting;
+import io.crate.data.Input;
+import io.crate.execution.engine.aggregation.AggregationFunction;
+import io.crate.execution.engine.aggregation.DocValueAggregator;
+import io.crate.memory.MemoryManager;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.ByteType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import io.crate.types.DoubleType;
+import io.crate.types.FloatType;
+import io.crate.types.IntegerType;
+import io.crate.types.LongType;
+import io.crate.types.ShortType;
+import org.apache.lucene.index.DocValues;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.util.NumericUtils;
+import org.elasticsearch.Version;
+import org.elasticsearch.common.breaker.CircuitBreakingException;
+import org.elasticsearch.index.mapper.MappedFieldType;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.List;
+
+public class NumericSumAggregation extends AggregationFunction<BigDecimal, BigDecimal> {
+
+    public static final String NAME = "sum";
+    public static final Signature SIGNATURE = Signature.aggregate(
+        NAME,
+        DataTypes.NUMERIC.getTypeSignature(),
+        DataTypes.NUMERIC.getTypeSignature()
+    );
+
+    public static void register(AggregationImplModule mod) {
+        mod.register(SIGNATURE, NumericSumAggregation::new);
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+    private final DataType<BigDecimal> returnType;
+
+    @VisibleForTesting
+    private NumericSumAggregation(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+
+        // usually the sum numeric aggregation function is going to be
+        // resolved after casting the sum parameter to the numeric type,
+        // e.g. sum(count::numeric(16, 2)). Having that, it
+        // is not possible to resolve the return type that will contain
+        // the precision and/or scale. Therefore, we infer the return
+        // type from the argument type.
```

Comment:
Shouldn't the returnType default to `numeric` anyway because of the signature definition?
Not sure I understand the reasoning here.

If we later on add supports for numeric type storage we shouldn't have to adjust aggregation functions.

Reply:
Yes, the return type defaults to `numeric`, but without default precision and 0 scale.  If we do smth like `sum(count::numeric(16, 2))` we will lose the precision and scale information in the return type, so currently that is the only way I think, otherwise, we would need to have use type variables instead of numeric in the sum signature and somehow make it work also with other types.